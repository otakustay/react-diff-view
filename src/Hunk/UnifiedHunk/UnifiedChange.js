/* eslint-disable no-empty-function */
import {memo, useMemo} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {mapValues} from 'lodash';
import {computeOldLineNumber, computeNewLineNumber} from '../../utils';
import CodeCell from '../CodeCell';
import '../Change.css';

const useBoundCallbacks = (callbacks, arg) => useMemo(
    () => mapValues(callbacks, fn => () => fn(arg)),
    [callbacks, arg]
);

const GutterCell = ({hide, className, lineNumber, gutterAnchor, anchorID, ...props}) => {
    if (hide) {
        return null;
    }

    return (
        <td className={className} {...props}>
            {
                gutterAnchor
                    ? <a href={'#' + anchorID} data-line-number={lineNumber}>{lineNumber}</a>
                    : lineNumber
            }
        </td>
    );
};

const UnifiedChange = props => {
    const {
        change,
        selected,
        tokens,
        className,
        gutterClassName,
        codeClassName,
        gutterEvents,
        codeEvents,
        hideGutter,
        gutterAnchor,
        generateAnchorID,
        renderToken,
    } = props;
    const {type, content} = change;
    const oldLine = computeOldLineNumber(change);
    const oldLineNumber = oldLine === -1 ? undefined : oldLine;
    const newLine = computeNewLineNumber(change);
    const newLineNumber = newLine === -1 ? undefined : newLine;

    const eventArg = useMemo(() => ({change}), [change]);
    const boundGutterEvents = useBoundCallbacks(gutterEvents, eventArg);
    const boundCodeEvents = useBoundCallbacks(codeEvents, eventArg);

    const anchorID = generateAnchorID(change);
    const gutterClassNameValue = classNames(
        'diff-gutter',
        `diff-gutter-${type}`,
        gutterClassName,
        {'diff-gutter-selected': selected}
    );
    const codeClassNameValue = classNames(
        'diff-code',
        `diff-code-${type}`,
        codeClassName,
        {'diff-code-selected': selected}
    );

    return (
        <tr
            id={anchorID}
            className={classNames('diff-line', className)}
        >
            <GutterCell
                hide={hideGutter}
                className={gutterClassNameValue}
                lineNumber={oldLineNumber}
                gutterAnchor={gutterAnchor}
                anchorID={anchorID}
                {...boundGutterEvents}
            />
            <GutterCell
                hide={hideGutter}
                className={gutterClassNameValue}
                lineNumber={newLineNumber}
                gutterAnchor={gutterAnchor}
                anchorID={anchorID}
                {...boundGutterEvents}
            />
            <CodeCell
                className={codeClassNameValue}
                text={content}
                tokens={tokens}
                renderToken={renderToken}
                {...boundCodeEvents}
            />
        </tr>
    );
};


UnifiedChange.propTypes = {
    selected: PropTypes.bool.isRequired,
    tokens: PropTypes.arrayOf(PropTypes.object),
};

UnifiedChange.defaultProps = {
    tokens: null,
};

export default memo(UnifiedChange);
