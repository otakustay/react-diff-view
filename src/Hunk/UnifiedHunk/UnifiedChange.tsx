// @ts-nocheck ignore: third-party
/* eslint-disable no-empty-function */
import {memo, useState, useMemo, useCallback} from 'react';
import classNames from 'classnames';
import {mapValues} from 'lodash';
import CodeCell from '../CodeCell';
import {composeCallback, renderDefaultBy, wrapInAnchorBy} from '../utils';

const useBoundCallbacks = (callbacks, arg, hoverOn, hoverOff) => useMemo(
    () => {
        const output = mapValues(callbacks, fn => () => fn(arg));
        output.onMouseEnter = composeCallback(hoverOn, output.onMouseEnter);
        output.onMouseLeave = composeCallback(hoverOff, output.onMouseLeave);
        return output;
    },
    [callbacks, hoverOn, hoverOff, arg]
);

const useBoolean = () => {
    const [value, setValue] = useState(false);
    const on = useCallback(
        () => setValue(true),
        []
    );
    const off = useCallback(
        () => setValue(false),
        []
    );
    return [value, on, off];
};

const renderGutterCell = (className, change, side, gutterAnchor, anchorTarget, events, inHoverState, renderGutter) => {
    const gutterOptions = {
        change,
        side,
        inHoverState,
        renderDefault: renderDefaultBy(change, side),
        wrapInAnchor: wrapInAnchorBy(gutterAnchor, anchorTarget),
    };

    return (
        <td className={className} {...events}>
            {renderGutter(gutterOptions)}
        </td>
    );
};

function UnifiedChange(props) {
    const {
        change,
        selected,
        tokens = null,
        className,
        gutterClassName,
        codeClassName,
        gutterEvents,
        codeEvents,
        hideGutter,
        gutterAnchor,
        generateAnchorID,
        renderToken,
        renderGutter,
    } = props;
    const {type, content} = change;

    const [hover, hoverOn, hoverOff] = useBoolean();
    const eventArg = useMemo(
        () => ({change}),
        [change]
    );
    const boundGutterEvents = useBoundCallbacks(gutterEvents, eventArg, hoverOn, hoverOff);
    const boundCodeEvents = useBoundCallbacks(codeEvents, eventArg, hoverOn, hoverOff);

    const anchorID = generateAnchorID(change);
    const gutterClassNameValue = classNames(
        'diff-gutter',
        `diff-gutter-${type}`,
        gutterClassName,
        {'diff-gutter-selected': selected}
    );
    const codeClassNameValue = classNames(
        'diff-code',
        `diff-code-${type}`,
        codeClassName,
        {'diff-code-selected': selected}
    );

    return (
        <tr id={anchorID} className={classNames('diff-line', className)}>
            {
                !hideGutter && renderGutterCell(
                    gutterClassNameValue,
                    change,
                    'old',
                    gutterAnchor,
                    anchorID,
                    boundGutterEvents,
                    hover,
                    renderGutter
                )
            }
            {
                !hideGutter && renderGutterCell(
                    gutterClassNameValue,
                    change,
                    'new',
                    gutterAnchor,
                    anchorID,
                    boundGutterEvents,
                    hover,
                    renderGutter
                )
            }
            <CodeCell
                className={codeClassNameValue}
                text={content}
                tokens={tokens}
                renderToken={renderToken}
                {...boundCodeEvents}
            />
        </tr>
    );
}

export default memo(UnifiedChange);
