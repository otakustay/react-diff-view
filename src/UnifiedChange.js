import {PureComponent} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {computeOldLineNumber, computeNewLineNumber} from './utils';
import {changePropType, eventsPropType, classNamesPropType} from './propTypes';
import {createEventsBindingSelector} from './selectors';
import './Change.css';

export default class UnifiedChange extends PureComponent {


    bindGutterEvents = createEventsBindingSelector();

    bindCodeEvents = createEventsBindingSelector();

    static propTypes = {
        change: changePropType.isRequired,
        selected: PropTypes.bool.isRequired,
        customEvents: eventsPropType,
        customClassNames: classNamesPropType,
        onRenderCode: PropTypes.func
    };

    static defaultProps = {
        customEvents: {},
        customClassNames: {},
        onRenderCode() {
        }
    };

    componentDidMount() {
        const {change, onRenderCode} = this.props;
        const cell = this.container.querySelector('.diff-code');
        onRenderCode(cell, change);
    }

    componentDidUpdate() {
        // TODO: 如何判断一个td里的代码是不是已经高亮过了？
    }

    render() {
        const {change, selected, customClassNames, customEvents, gutterAnchor, generateAnchorID} = this.props;
        const {type, content} = change;
        const oldLine = computeOldLineNumber(change);
        const oldLineNumber = oldLine === -1 ? undefined : oldLine;
        const newLine = computeNewLineNumber(change);
        const newLineNumber = newLine === -1 ? undefined : newLine;

        const boundGutterEvents = this.bindGutterEvents(customEvents.gutter, change);
        const boundCodeEvents = this.bindCodeEvents(customEvents.code, change);

        const anchorID = generateAnchorID(change);
        const gutterClassName = classNames(
            'diff-gutter',
            `diff-gutter-${type}`,
            customClassNames.gutter,
            {'diff-gutter-selected': selected}
        );
        const codeClassName = classNames(
            'diff-code',
            `diff-code-${type}`,
            customClassNames.code,
            {'diff-code-selected': selected}
        );

        return (
            <tr
                id={anchorID}
                className={classNames('diff-line', customClassNames.line)}
                ref={container => (this.container = container)}
            >
                <td className={gutterClassName} data-line-number={oldLineNumber} {...boundGutterEvents}>
                    {gutterAnchor ? <a href={'#' + anchorID} data-line-number={oldLineNumber} /> : null}
                </td>
                <td className={gutterClassName} data-line-number={newLineNumber} {...boundGutterEvents}>
                    {gutterAnchor ? <a href={'#' + anchorID} data-line-number={newLineNumber} /> : null}
                </td>
                <td className={codeClassName} {...boundCodeEvents}>{content}</td>
            </tr>
        );
    }
}
