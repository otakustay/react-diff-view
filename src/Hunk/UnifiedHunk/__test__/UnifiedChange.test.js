import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {partialRight, get} from 'lodash';
import UnifiedChange from '../UnifiedChange';

const defaultChange = {
    content: 'text',
};

const pickId = partialRight(get, 'content');

describe('UnifiedChange', () => {
    test('renders correctly', () => {
        expect(renderer.create(
            <UnifiedChange change={defaultChange} generateAnchorID={pickId} />
        ).toJSON()).toMatchSnapshot();
    });

    test('gutterEvents should be called with side', () => {
        const onClick = jest.fn();
        const element = (
            <UnifiedChange
                selected={false}
                change={defaultChange}
                generateAnchorID={pickId}
                gutterEvents={{onClick}}
            />
        );
        const wrapper = shallow(element);
        wrapper.find('.diff-gutter').at(1).simulate('click');
        // TODO it should be called with {change: defaultChange, side: 'new'}
        expect(onClick).toBeCalledWith({change: defaultChange, side: undefined});
    });
});
