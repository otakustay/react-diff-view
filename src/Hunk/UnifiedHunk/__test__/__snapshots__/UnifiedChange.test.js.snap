// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnifiedChange renders correctly 1`] = `
<tr
  className="diff-line"
  id="text"
>
  <td
    className="diff-gutter diff-gutter-undefined"
  />
  <td
    className="diff-gutter diff-gutter-undefined"
  />
  <td
    className="diff-code diff-code-undefined"
  >
    text
  </td>
</tr>
`;
